src/System.Linq/src/System/Linq/DefaultIfEmpty.cs

```diff
@@ -17,24 +19,95 @@ public static IEnumerable<TSource> DefaultIfEmpty<TSource>(this IEnumerable<TSou
         public static IEnumerable<TSource> DefaultIfEmpty<TSource>(this IEnumerable<TSource> source, TSource defaultValue)
```

Comment:
It certainly is very useful in the outer join case, but not so much beyond that. In terms of implementation details it's a prime candidate for implementing IPartition, but in terms of use it wouldn't justify the type check.